### PR TITLE
Rounded corners on kirby-list (issue #160)

### PR DESCRIPTION
**Pull request by askarby**
With this PR, we're now able to render a kirby-list with rounded corners (defaults to "no rounded corners", which makes it backwards compatible).

To implement this, I've done the following the following rules:
- An `@Input` parameter is able to supply a shape (`square` renders the list the way it used to be rendered - ie. the backwards-compatibility mode, automatically applied when binding to `shape`-input  is omitted or `rounded` when rounded corners should be rendered.
- The shape-parameter is applied as a class to each row (eg. `.square` or `.rounded`).
- The logic for rendering with "rounded corners" can be summarized very briefly! If the list is **without** sections, the first and last rows (of the list) are applied border radii ("top left" and "top right" for first row, and "bottom left" and "bottom right" for last row). If the list is **with** sections, the first  and last rows (of the sections) are applied rounded corners.
- The indication of which row is "first" and "last" is conveyed by applying the class "top" for the first row, and "bottom" for the last row (maybe `first` and `last` would be better names for those classes?)
- Calculating which rows are first and last are done "up front" (eg. when items input is changed), to save computational power (the `getClasses`-method that is bound to ngClass-directive and that is triggered very frequently, hence should be as light-weight as possible to invoke).
- I've tested on web and {N}-environments, and running with 6000 rows (on native) seems to work like a charm (no noticable delay). Web becomes unresponsive, but I won't optimize on this, since ionic should fix that for us (fingers crossed) - in short, we need a mechanism for "virtual scrolling" the list, since the browser dies on reflowing several thousands DOM-node insertions.
- Lastly, a few tweaks have been made to border styling, to avoid double borders, and actually drawing meaningful borders when `rounded`-shape is enable (only drawing the top-border does not render anything meaning, so changed that).

I'm open for suggestions, improvements, criticism etc.

### CODE REVIEW

File: src/kirby/components/list/list.component.html [hunk 1]
```diff
@@ -15,7 +15,7 @@
           ">
         </ng-container>
       </li>
-      <li class="row" [ngClass]="{'selectable': isSelectable}" *ngFor="let item of section.items" (click)="onItemSelect(item)" tabindex="0" [attr.role]="isSelectable ? 'button' : null">
+      <li class="row" [ngClass]="getClasses(item)" *ngFor="let item of section.items" (click)="onItemSelect(item)" tabindex="0" [attr.role]="isSelectable ? 'button' : null">
```
Comment (jakobe) [inline src/kirby/components/list/list.component.html]:
```suggestion
      <li class="row" [ngClass]="getCssClasses(item)" *ngFor="let item of section.items" (click)="onItemSelect(item)" tabindex="0" [attr.role]="isSelectable ? 'button' : null">
```

File: src/kirby/components/list/list.component.html [hunk 2]
```diff
@@ -31,7 +31,7 @@
   </ng-container>
 
   <ng-template #normalList>
-    <li class="row" [ngClass]="{'selectable': isSelectable}" *ngFor="let item of items" (click)="onItemClick(item)" tabindex="0" [attr.role]="isSelectable ? 'button' : null">
+    <li class="row" [ngClass]="getClasses(item)" *ngFor="let item of items" (click)="onItemSelect(item)" tabindex="0" [attr.role]="isSelectable ? 'button' : null">
```
Comment (jakobe) [inline src/kirby/components/list/list.component.html]:
```suggestion
    <li class="row" [ngClass]="getCssClasses(item)" *ngFor="let item of items" (click)="onItemSelect(item)" tabindex="0" [attr.role]="isSelectable ? 'button' : null">
```

File: src/kirby/components/list/list.component.tns.html [hunk 3]
```diff
@@ -9,7 +9,7 @@
   >
     <!-- ITEM -->
     <ng-template tkListItemTemplate let-item="item">
-      <FlexboxLayout class="row" (tap)="onItemSelect(item)">
+      <FlexboxLayout class="row" [ngClass]="getClasses(item)" (tap)="onItemSelect(item)">
```
Comment (jakobe) [inline src/kirby/components/list/list.component.tns.html]:
```suggestion
      <FlexboxLayout class="row" [ngClass]="getCssClasses(item)" (tap)="onItemSelect(item)">
```

File: src/kirby/components/list/list.component.ts [hunk 4]
```diff
@@ -66,4 +101,65 @@ export class ListComponent implements OnInit {
   onLoadOnDemand(event?: LoadOnDemandEventData) {
     this.listHelper.onLoadOnDemand(this, event);
   }
+
+  getClasses(current: any) {
```
Comment (jakobe) [inline src/kirby/components/list/list.component.ts]: :fire: Move this entire section (`getClasses`, `determineClasses`, `getClassesForItem`) to helper/utility class and only apply for `.tns` (web has this build into css pseudo selectors)

File: src/kirby/components/list/list.component.scss [hunk 5]
```diff
@@ -17,19 +16,33 @@ ul {
...
-      &:first-of-type {
-        border-top: 1px solid get-color('kirby-grey-2');
+      border-color: get-color('kirby-grey-2');
+      border-style: solid;
+      border-width: 0 0 1px 0;
+
+      &.rounded {
+        border-width: 0 1px 1px 1px;
+
+        &.top {
```
Comment (jakobe) [inline src/kirby/components/list/list.component.scss]: :fire: web already has `top`/`bottom` build into css pseudo selectors (`first-child`/`last-child`, `first-of-type` etc), so it shouldn't be neccesary to implement lowest common denominator (aka {N} ;) )  for web

File: src/kirby/components/list/list.component.ts (hunk 4 above)
Comment (askarby) [reply src/kirby/components/list/list.component.ts]: Maybe it's just me, but I can't see how any CSS-(pseudo)selector can be applied to determine "top-most" row of a section, when the "rows" and "sections headers" are all rendered in a flat li-structure?

File: src/kirby/components/list/list.component.scss (hunk 5 above)
Comment (askarby) [reply src/kirby/components/list/list.component.scss]: Please see the comment I made on `src/kirby/components/list/list.component.ts`

File: package.json [hunk 6]
```diff
@@ -57,6 +57,7 @@
     "nativescript-webview-interface": "^1.4.2",
     "prismjs": "^1.15.0",
     "reflect-metadata": "~0.1.8",
+    "reveal-plugin-ios": "file:../reveal-plugin-ios",
```
Comment (jakobe) [inline package.json]: Remove this

File: package.json [hunk 7]
```diff
@@ -66,6 +67,7 @@
     "@angular/cli": "^7.3.1",
     "@angular/compiler-cli": "^7.2.4",
     "@nativescript/schematics": "^0.5.0",
+    "@netbasal/spectator": "^3.6.1",
```
Comment (jakobe) [inline package.json]: This is something we want/need?

File: src/app/app.module.ts [hunk 8]
```diff
@@ -10,10 +10,11 @@ import { SideNavComponent } from './page/side-nav/side-nav.component';
 import { HeaderComponent } from './page/header/header.component';
 import { IntroComponent } from './intro/intro.component';
 import { KirbyModule } from '../kirby/kirby.module';
+import { IonicModule } from '@ionic/angular';
 
 @NgModule({
   declarations: [AppComponent, HomeComponent, SideNavComponent, HeaderComponent, IntroComponent],
-  imports: [BrowserModule, AppRoutingModule, HttpClientModule, FormsModule, KirbyModule],
+    imports: [BrowserModule, AppRoutingModule, HttpClientModule, FormsModule, KirbyModule, IonicModule],
```
Comment (jakobe) [inline src/app/app.module.ts]: `, IonicModule.forRoot()],`?

File: src/app/examples/examples.module.tns.ts [hunk 9]
```diff
@@ -2,14 +2,15 @@
...
+import { NativeScriptFormsModule } from 'nativescript-angular';
 
 import { KirbyModule } from '../../kirby/kirby.module';
 import { COMPONENT_DECLARATIONS } from './examples.common';
 import { NativeScriptDoughnutChartExampleComponent } from './nativescript-only/doughnut-chart-example/doughnut-chart-example.component.tns-only';
 import { NativeScriptLineChartExampleComponent } from './nativescript-only/line-chart-example/line-chart-example.component.tns-only';
-import { ListSectionExampleComponent } from './list/tns-list-examples/list-section-example/list-section-example.component.tns';
-import { ListCustomCellLinesExampleComponent } from './list/tns-list-examples/list-custom-cell-lines-example/list-custom-cell-lines-example.component.tns';
-import { ListItemExampleComponent } from './list/tns-list-examples/list-item-example/list-item-example.component.tns';
+import { ListItemExampleComponent } from '~/app/examples/list/list-item-example/list-item-example.component.tns';
```
Comment (jakobe) [inline src/app/examples/examples.module.tns.ts]: Don't use absolute imports - it's easer to grasp that this lives inside the examples folder as does the module

File: src/app/examples/nativescript-only/list-section-example/list-section-example.component.tns.ts [hunk 10]
```diff
@@ -1,13 +1,20 @@
 import { Component } from '@angular/core';
 
-import { BaseListComponent } from '../../base-list.component';
+import { BaseListComponent } from '~/app/examples/list/base-list.component';
```
Comment (jakobe) [inline src/app/examples/nativescript-only/list-section-example/list-section-example.component.tns.ts]: Don't use absolute imports - it's easer to grasp that this lives inside the examples folder as does this component

File: src/app/examples/nativescript-only/list-section-example/list-section-example.component.tns.ts [hunk 11]
```diff
@@ -1,13 +1,20 @@
...
 })
 export class ListSectionExampleComponent extends BaseListComponent {
   getSectionName(item: any): string {
-    return item.detail > 0 ? 'Positive' : 'Negative';
+    if (item.detail > 0) {
+      return 'Positive';
+    } else if (item.detail < 0) {
+      return 'Negative';
+    } else {
+      return 'Nul';
```
Comment (jakobe) [inline src/app/examples/nativescript-only/list-section-example/list-section-example.component.tns.ts]: ??

File: src/app/showcase/list-showcase/list-showcase.component.tns.html [hunk 12]
```diff
@@ -1 +1 @@
-<Button text="list-showcase works!" class="btn btn-primary"></Button>
\ No newline at end of file
+<<Button text="list-showcase works!" class="btn btn-primary"></Button>>
```
Comment (jakobe) [inline src/app/showcase/list-showcase/list-showcase.component.tns.html]: Remove extra `<` `>`

File: src/kirby/components/list/list.component.scss [hunk 13]
```diff
@@ -2,11 +2,11 @@
 ul {
   list-style-type: none;
   padding-left: 0;
-  margin-top: 0;
+  margin: 0;
   font-size: font-size('s');
 
   li {
-    background-color: get-color('kirby-white');
+    border-bottom: 1px solid get-color('kirby-grey-2');
```
Comment (jakobe) [inline src/kirby/components/list/list.component.scss]: Shouldn't this only be enabled when `showDivider=true`?

File: src/kirby/components/list/list.component.ts [hunk 14]
```diff
@@ -20,14 +21,15 @@ import {
...
+import { ListRowClassService } from '~/kirby/components/list/list-row-class.service';
 
 export type ListShape = 'square' | 'rounded';
 
 @Component({
   selector: 'kirby-list',
   templateUrl: './list.component.html',
   styleUrls: ['./list.component.scss'],
-  providers: [ListHelper],
+  providers: [ListHelper, ListRowClassService],
```
Comment (jakobe) [inline src/kirby/components/list/list.component.ts]: Maybe `ListRowClassHelper`? Or merged into `ListHelper`?

File: package.json (hunk 7 above)
Comment (askarby) [reply package.json]: Yes, it makes testing component lifecycle hooks (requirements of host-components) much easier.

File: package.json (hunk 6 above)
Comment (askarby) [reply package.json]: Not sure how this got in here (wasn't on purpose)... will remove

File: src/app/app.module.ts (hunk 8 above)
Comment (askarby) [reply src/app/app.module.ts]: Fixed!

File: src/app/examples/examples.module.tns.ts (hunk 9 above)
Comment (askarby) [reply src/app/examples/examples.module.tns.ts]: Removed, and removed `paths`-element from `tsconfig.json`

File: src/app/examples/nativescript-only/list-section-example/list-section-example.component.tns.ts (hunk 10 above)
Comment (askarby) [reply src/app/examples/nativescript-only/list-section-example/list-section-example.component.tns.ts]: Fixed!

File: src/app/examples/nativescript-only/list-section-example/list-section-example.component.tns.ts (hunk 11 above)
Comment (askarby) [reply src/app/examples/nativescript-only/list-section-example/list-section-example.component.tns.ts]: Leftovers from testing variations of list... reverting back to previous example-implementation

File: src/app/showcase/list-showcase/list-showcase.component.tns.html (hunk 12 above)
Comment (askarby) [reply src/app/showcase/list-showcase/list-showcase.component.tns.html]: Yeah, that's just silly... removed!

File: src/kirby/components/list/list.component.scss (hunk 13 above)
Comment (askarby) [reply src/kirby/components/list/list.component.scss]: True... left-over styling remains (that wasn't removed)... will do so!

File: src/kirby/components/list/list.component.ts (hunk 14 above)
Comment (askarby) [reply src/kirby/components/list/list.component.ts]: I think the words "Helper" and "Utils" are the most abused (and horrific) names for constructs in software.

`ListHelper` says that it "helps the list"... but, to do what? I've (previously) experienced that a "Helper"-class becomes a box of arbitrary tools with no to little "low coupling, high cohesion" 

I'm not opposed to renaming `ListRowClassService` to something else, just not sure of a better alternative... any suggestions?

File: package.json (hunk 7 above)
Comment (jakobe) [reply package.json]: Ok

File: src/kirby/components/list/list.component.ts (hunk 14 above)
Comment (SilverGull) [reply src/kirby/components/list/list.component.ts]: Lets keep it as is. I cannot come up with anything better and it is a service for getting css classes for rows. The best reason for renaming would be to align naming the naming to match the ListHelper.

File: src/kirby/components/list/list.component.tns.html [hunk 15]
```diff
@@ -6,12 +6,17 @@
...
     <!-- ITEM -->
     <ng-template tkListItemTemplate let-item="item">
-      <FlexboxLayout class="row" [ngClass]="{'divider': showDivider}" (tap)="onItemSelect(item)">
+      <FlexboxLayout class="row"
+                     [class.divider]="showDivider"
+                     [ngClass]="getCssClasses(item)"
+                     (tap)="onItemSelect(item)">
         <ng-container
           *ngTemplateOutlet="listItemTemplate; context: { $implicit: item }">
+          <!--Label [text]="getCssClasses(item)"></Label-->
```
Comment (jakobe) [inline src/kirby/components/list/list.component.tns.html]: Delete 🔥 😄 
```suggestion
```

File: src/kirby/components/list/list.component.html [hunk 16]
```diff
@@ -1,7 +1,7 @@
 <!-- HEADER-->
-<ng-container *ngIf="listHeaderTemplate"> 
+<div class="list-header" *ngIf="listHeaderTemplate">
   <ng-container *ngTemplateOutlet="listHeaderTemplate" ></ng-container>
-</ng-container>
+</div>
 
 <!-- BODY-->
 <ul kirbyInfiniteScroll (scrollEnd)="onLoadOnDemand()" [disabled]="!isLoadOnDemandEnabled">
```
Comment (jakobe) [inline src/kirby/components/list/list.component.html]: Reuse groupedItems:
`<ng-container *ngFor="let section of groupedItems">`

File: src/app/examples/examples-routing.module.tns.ts [hunk 17]
```diff
@@ -7,10 +7,10 @@ import { routes } from './examples.routes';
 import { ExamplesModule } from './examples.module';
 import { NativeScriptDoughnutChartExampleComponent } from './nativescript-only/doughnut-chart-example/doughnut-chart-example.component.tns-only';
 import { NativeScriptLineChartExampleComponent } from './nativescript-only/line-chart-example/line-chart-example.component.tns-only';
-import { ListSectionExampleComponent } from './list/tns-list-examples/list-section-example/list-section-example.component.tns';
-import { ListCustomCellLinesExampleComponent } from './list/tns-list-examples/list-custom-cell-lines-example/list-custom-cell-lines-example.component.tns';
-import { ListItemExampleComponent } from './list/tns-list-examples/list-item-example/list-item-example.component.tns';
 import { ListLoadOnDemandExampleComponent } from './list/load-on-demand/list-load-on-demand-example.component';
+import { ListCustomCellLinesExampleComponent } from '~/app/examples/list/list-custom-cell-lines-example/list-custom-cell-lines-example.component.tns';
```
[the remark this answers is not in the document]
Comment (askarby) [reply src/app/examples/examples-routing.module.tns.ts]: Automated tooling at its worst... fixing it

File: src/app/examples/examples.module.tns.ts [hunk 18]
```diff
@@ -2,14 +2,15 @@
 
 import { NgModule, NO_ERRORS_SCHEMA } from '@angular/core';
 import { NativeScriptCommonModule } from 'nativescript-angular/common';
+import { NativeScriptFormsModule } from 'nativescript-angular';
```
[the remark this answers is not in the document]
Comment (askarby) [reply src/app/examples/examples.module.tns.ts]: I have no clue, helpful tool trying to be helpful, and failing at it!?

File: src/app/examples/examples.module.tns.ts [hunk 19]
```diff
@@ -23,7 +24,7 @@ const nativeScriptDeclarations = [
 ];
 
 @NgModule({
-  imports: [NativeScriptCommonModule, KirbyModule],
+  imports: [NativeScriptCommonModule, KirbyModule, NativeScriptFormsModule],
```
[the remark this answers is not in the document]
Comment (askarby) [reply src/app/examples/examples.module.tns.ts]: I'm pretty sure I got this by merging master into this feature branch, don't worry... will fix it.

File: src/app/examples/examples.module.tns.ts (hunk 18 above)
[the remark this answers is not in the document]
Comment (askarby) [reply src/app/examples/examples.module.tns.ts]: I think this is an effect of merging master to feature branch, leaving it.

File: src/kirby/components/list/list.component.tns.html (hunk 15 above)
Comment (askarby) [reply src/kirby/components/list/list.component.tns.html]: Agreed!

File: src/kirby/components/list/list.component.html (hunk 16 above)
Comment (askarby) [reply src/kirby/components/list/list.component.html]: Agreed!

File: src/kirby/components/list/list.component.scss [hunk 20]
```diff
@@ -17,28 +16,75 @@ ul {
...
+
+      &.divider:not(:last-child) {
+        border-bottom-width: 1px;
+      }
 
-    &.section-header {
-      background-color: transparent;
-      border: none;
+      &.selectable:hover {
+        background-color: get-color('kirby-secondary');
```
[the remark this answers is not in the document]
Comment (askarby) [reply src/kirby/components/list/list.component.scss]: Changed to `kirby-base-color`, which now clashes with background-color of the kirby cookbook examples, but it at least doesn't have as much constrast as with the `kirby-secondary`.

File: src/app/examples/examples.module.tns.ts (hunk 18 above)
[the remark this answers is not in the document]
Comment (askarby) [reply src/app/examples/examples.module.tns.ts]: Removed the import, application doesn't seem to use it!

File: src/app/examples/examples.module.tns.ts (hunk 19 above)
[the remark this answers is not in the document]
Comment (askarby) [reply src/app/examples/examples.module.tns.ts]: Removed import and dependency, didn't seem to be used by application.

File: src/app/examples/nativescript-only/list-section-example/list-section-example.component.tns.ts (hunk 10 above)
Comment (askarby) [reply src/app/examples/nativescript-only/list-section-example/list-section-example.component.tns.ts]: Reverted back to relative import

File: src/app/app.module.ts [hunk 21]
```diff
@@ -13,7 +14,14 @@ import { KirbyModule } from '../kirby/kirby.module';
...
 @NgModule({
   declarations: [AppComponent, HomeComponent, SideNavComponent, HeaderComponent, IntroComponent],
-  imports: [BrowserModule, AppRoutingModule, HttpClientModule, FormsModule, KirbyModule],
+  imports: [
+    BrowserModule,
+    AppRoutingModule,
+    HttpClientModule,
+    FormsModule,
+    KirbyModule,
+    IonicModule.forRoot(),
```
[the remark this answers is not in the document]
Comment (askarby) [reply src/app/app.module.ts]: No clue, reverted!

**Comment by jakobe** [comment]
Closes #160

**Pull request merged**